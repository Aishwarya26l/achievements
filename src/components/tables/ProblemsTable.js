/**
 * @file ProblemsTable component module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 22.02.18
 */

import React from "react";
import PropTypes from "prop-types";

import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";

import Button from "@material-ui/core/Button";

import { Link } from "react-router-dom";

import withStyles from "@material-ui/core/styles/withStyles";
import { pathProblemDialogShow } from "../../containers/Paths/actions";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class ProblemsTable extends React.PureComponent {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    currentUserId: PropTypes.string,
    dispatch: PropTypes.func.isRequired,
    pathOwnerId: PropTypes.any,
    problems: PropTypes.array.isRequired,
    selectedPathId: PropTypes.string
  };

  onEditClick = problem => this.props.dispatch(pathProblemDialogShow(problem));

  render() {
    const {
      classes,
      currentUserId,
      pathOwnerId,
      problems,
      selectedPathId
    } = this.props;

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Problem name</TableCell>
            {currentUserId !== pathOwnerId && <TableCell>Progress</TableCell>}
            <TableCell
              style={{
                width: 240
              }}
            >
              Actions
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {problems.map(problem => (
            <TableRow key={problem.id}>
              <TableCell>{problem.name}</TableCell>
              {currentUserId !== pathOwnerId && (
                <TableCell>{problem.solved}</TableCell>
              )}
              <TableCell>
                <Link
                  className={classes.link}
                  to={`/paths/${selectedPathId || pathOwnerId}/problems/${
                    problem.id
                  }`}
                >
                  <Button variant="raised">Open</Button>
                </Link>
                {pathOwnerId === currentUserId && (
                  <Button
                    className={classes.button}
                    onClick={() => this.onEditClick(problem)}
                    variant="raised"
                  >
                    Edit
                  </Button>
                )}
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(ProblemsTable);
