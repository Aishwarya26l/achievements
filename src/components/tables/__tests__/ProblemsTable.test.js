import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import ProblemsTable from "../ProblemsTable";
import Button from "@material-ui/core/Button";
import sinon from "sinon";

describe("<ProblemsTable>", () => {
  let shallow;
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow({
      dive: true
    });
  });

  it("should generate row", () => {
    const wrapper = shallow(
      <ProblemsTable
        dispatch={mockDispatch}
        problems={[
          {
            id: "test",
            name: "Test"
          }
        ]}
      />
    );

    wrapper
      .find(Button)
      .at(1)
      .simulate("click");
    expect(
      mockDispatch.calledWith({
        type: "PATH_PROBLEM_DIALOG_SHOW",
        problemInfo: {
          id: "test",
          name: "Test"
        }
      })
    ).toEqual(true);
  });
});
