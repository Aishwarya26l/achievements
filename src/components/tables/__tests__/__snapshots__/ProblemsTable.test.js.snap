// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProblemsTable> should check snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProblemsTable
    classes={
        Object {
            "button": "ProblemsTable-button-2",
            "link": "ProblemsTable-link-1",
          }
    }
    dispatch={[Function]}
    problems={
        Array [
            Object {
              "id": "test",
              "name": "Test",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(TableHead)>
          <WithStyles(TableRow)>
                    <WithStyles(TableCell)>
                              Problem name
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)
                              style={
                                        Object {
                                                  "width": 240,
                                                }
                              }
                    >
                              Actions
                    </WithStyles(TableCell)>
          </WithStyles(TableRow)>
</WithStyles(TableHead)>,
        <WithStyles(TableBody)>
          <WithStyles(TableRow)>
                    <WithStyles(TableCell)>
                              Test
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                              <Link
                                        className="ProblemsTable-link-1"
                                        replace={false}
                                        to="/paths/undefined/problems/test"
                              >
                                        <WithStyles(Button)
                                                  variant="raised"
                                        >
                                                  Open
                                        </WithStyles(Button)>
                              </Link>
                              <WithStyles(Button)
                                        className="ProblemsTable-button-2"
                                        onClick={[Function]}
                                        variant="raised"
                              >
                                        Edit
                              </WithStyles(Button)>
                    </WithStyles(TableCell)>
          </WithStyles(TableRow)>
</WithStyles(TableBody)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <WithStyles(TableRow)>
            <WithStyles(TableCell)>
                        Problem name
            </WithStyles(TableCell)>
            <WithStyles(TableCell)
                        style={
                                    Object {
                                                "width": 240,
                                              }
                        }
            >
                        Actions
            </WithStyles(TableCell)>
</WithStyles(TableRow)>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(TableCell)>
                Problem name
</WithStyles(TableCell)>,
              <WithStyles(TableCell)
                style={
                                Object {
                                                "width": 240,
                                              }
                }
>
                Actions
</WithStyles(TableCell)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Problem name",
              },
              "ref": null,
              "rendered": "Problem name",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Actions",
                "style": Object {
                  "width": 240,
                },
              },
              "ref": null,
              "rendered": "Actions",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(TableRow)>
              <WithStyles(TableCell)>
                            Test
              </WithStyles(TableCell)>
              <WithStyles(TableCell)>
                            <Link
                                          className="ProblemsTable-link-1"
                                          replace={false}
                                          to="/paths/undefined/problems/test"
                            >
                                          <WithStyles(Button)
                                                        variant="raised"
                                          >
                                                        Open
                                          </WithStyles(Button)>
                            </Link>
                            <WithStyles(Button)
                                          className="ProblemsTable-button-2"
                                          onClick={[Function]}
                                          variant="raised"
                            >
                                          Edit
                            </WithStyles(Button)>
              </WithStyles(TableCell)>
</WithStyles(TableRow)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "test",
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <WithStyles(TableCell)>
                  Test
</WithStyles(TableCell)>,
                <WithStyles(TableCell)>
                  <Link
                                    className="ProblemsTable-link-1"
                                    replace={false}
                                    to="/paths/undefined/problems/test"
                  >
                                    <WithStyles(Button)
                                                      variant="raised"
                                    >
                                                      Open
                                    </WithStyles(Button)>
                  </Link>
                  <WithStyles(Button)
                                    className="ProblemsTable-button-2"
                                    onClick={[Function]}
                                    variant="raised"
                  >
                                    Edit
                  </WithStyles(Button)>
</WithStyles(TableCell)>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Test",
                },
                "ref": null,
                "rendered": "Test",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    <Link
                      className="ProblemsTable-link-1"
                      replace={false}
                      to="/paths/undefined/problems/test"
>
                      <WithStyles(Button)
                                            variant="raised"
                      >
                                            Open
                      </WithStyles(Button)>
</Link>,
                    <WithStyles(Button)
                      className="ProblemsTable-button-2"
                      onClick={[Function]}
                      variant="raised"
>
                      Edit
</WithStyles(Button)>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": <WithStyles(Button)
                        variant="raised"
>
                        Open
</WithStyles(Button)>,
                      "className": "ProblemsTable-link-1",
                      "replace": false,
                      "to": "/paths/undefined/problems/test",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Open",
                        "variant": "raised",
                      },
                      "ref": null,
                      "rendered": "Open",
                      "type": [Function],
                    },
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Edit",
                      "className": "ProblemsTable-button-2",
                      "onClick": [Function],
                      "variant": "raised",
                    },
                    "ref": null,
                    "rendered": "Edit",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(TableHead)>
            <WithStyles(TableRow)>
                        <WithStyles(TableCell)>
                                    Problem name
                        </WithStyles(TableCell)>
                        <WithStyles(TableCell)
                                    style={
                                                Object {
                                                            "width": 240,
                                                          }
                                    }
                        >
                                    Actions
                        </WithStyles(TableCell)>
            </WithStyles(TableRow)>
</WithStyles(TableHead)>,
          <WithStyles(TableBody)>
            <WithStyles(TableRow)>
                        <WithStyles(TableCell)>
                                    Test
                        </WithStyles(TableCell)>
                        <WithStyles(TableCell)>
                                    <Link
                                                className="ProblemsTable-link-1"
                                                replace={false}
                                                to="/paths/undefined/problems/test"
                                    >
                                                <WithStyles(Button)
                                                            variant="raised"
                                                >
                                                            Open
                                                </WithStyles(Button)>
                                    </Link>
                                    <WithStyles(Button)
                                                className="ProblemsTable-button-2"
                                                onClick={[Function]}
                                                variant="raised"
                                    >
                                                Edit
                                    </WithStyles(Button)>
                        </WithStyles(TableCell)>
            </WithStyles(TableRow)>
</WithStyles(TableBody)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(TableRow)>
              <WithStyles(TableCell)>
                            Problem name
              </WithStyles(TableCell)>
              <WithStyles(TableCell)
                            style={
                                          Object {
                                                        "width": 240,
                                                      }
                            }
              >
                            Actions
              </WithStyles(TableCell)>
</WithStyles(TableRow)>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <WithStyles(TableCell)>
                  Problem name
</WithStyles(TableCell)>,
                <WithStyles(TableCell)
                  style={
                                    Object {
                                                      "width": 240,
                                                    }
                  }
>
                  Actions
</WithStyles(TableCell)>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Problem name",
                },
                "ref": null,
                "rendered": "Problem name",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Actions",
                  "style": Object {
                    "width": 240,
                  },
                },
                "ref": null,
                "rendered": "Actions",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(TableRow)>
                <WithStyles(TableCell)>
                                Test
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                                <Link
                                                className="ProblemsTable-link-1"
                                                replace={false}
                                                to="/paths/undefined/problems/test"
                                >
                                                <WithStyles(Button)
                                                                variant="raised"
                                                >
                                                                Open
                                                </WithStyles(Button)>
                                </Link>
                                <WithStyles(Button)
                                                className="ProblemsTable-button-2"
                                                onClick={[Function]}
                                                variant="raised"
                                >
                                                Edit
                                </WithStyles(Button)>
                </WithStyles(TableCell)>
</WithStyles(TableRow)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "test",
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <WithStyles(TableCell)>
                    Test
</WithStyles(TableCell)>,
                  <WithStyles(TableCell)>
                    <Link
                                        className="ProblemsTable-link-1"
                                        replace={false}
                                        to="/paths/undefined/problems/test"
                    >
                                        <WithStyles(Button)
                                                            variant="raised"
                                        >
                                                            Open
                                        </WithStyles(Button)>
                    </Link>
                    <WithStyles(Button)
                                        className="ProblemsTable-button-2"
                                        onClick={[Function]}
                                        variant="raised"
                    >
                                        Edit
                    </WithStyles(Button)>
</WithStyles(TableCell)>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Test",
                  },
                  "ref": null,
                  "rendered": "Test",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": Array [
                      <Link
                        className="ProblemsTable-link-1"
                        replace={false}
                        to="/paths/undefined/problems/test"
>
                        <WithStyles(Button)
                                                variant="raised"
                        >
                                                Open
                        </WithStyles(Button)>
</Link>,
                      <WithStyles(Button)
                        className="ProblemsTable-button-2"
                        onClick={[Function]}
                        variant="raised"
>
                        Edit
</WithStyles(Button)>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": <WithStyles(Button)
                          variant="raised"
>
                          Open
</WithStyles(Button)>,
                        "className": "ProblemsTable-link-1",
                        "replace": false,
                        "to": "/paths/undefined/problems/test",
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "class",
                        "props": Object {
                          "children": "Open",
                          "variant": "raised",
                        },
                        "ref": null,
                        "rendered": "Open",
                        "type": [Function],
                      },
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Edit",
                        "className": "ProblemsTable-button-2",
                        "onClick": [Function],
                        "variant": "raised",
                      },
                      "ref": null,
                      "rendered": "Edit",
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
